File: services/news_service.py
"""
News Service for Weekly Gold Market News
سرویس اخبار برای اخبار هفتگی بازار طلا
"""

import logging
import requests
from datetime import datetime, timedelta
from typing import List, Dict, Any, Optional
import re
from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)

class NewsService:
    """Service for fetching and processing gold market news"""
    
    def __init__(self):
        self.news_sources = {
            'forex_factory': 'https://www.forexfactory.com/calendar',
            'investing_gold': 'https://www.investing.com/commodities/gold-news',
            'marketwatch_gold': 'https://www.marketwatch.com/investing/future/gc00',
            'fxstreet_gold': 'https://www.fxstreet.com/commodities/gold'
        }
        
        self.headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
            'Accept-Language': 'en-US,en;q=0.5',
            'Accept-Encoding': 'gzip, deflate',
            'Connection': 'keep-alive',
        }

    def fetch_news_from_source(self, source_name: str, url: str, limit: int = 5) -> List[Dict[str, Any]]:
        """Fetch news from a specific source"""
        try:
            response = requests.get(url, headers=self.headers, timeout=30)
            response.raise_for_status()
            
            soup = BeautifulSoup(response.content, 'html.parser')
            news_items = []
            
            # Different parsing logic for each source
            if source_name == 'forex_factory':
                news_items = self._parse_forex_factory(soup, limit)
            elif source_name == 'investing_gold':
                news_items = self._parse_investing_com(soup, limit)
            elif source_name == 'marketwatch_gold':
                news_items = self._parse_marketwatch(soup, limit)
            elif source_name == 'fxstreet_gold':
                news_items = self._parse_fxstreet(soup, limit)
            
            return news_items
            
        except Exception as e:
            logger.error(f"Error fetching news from {source_name}: {e}")
            return []

    def _parse_forex_factory(self, soup: BeautifulSoup, limit: int) -> List[Dict[str, Any]]:
        """Parse Forex Factory news"""
        news_items = []
        try:
            # Look for news items related to gold, USD, inflation, etc.
            news_elements = soup.find_all(['div', 'tr'], class_=re.compile(r'calendar|news|event'))
            
            for element in news_elements[:limit]:
                title_elem = element.find(['span', 'a', 'td'], string=re.compile(r'gold|Gold|USD|inflation|Fed|FOMC', re.I))
                if title_elem:
                    title = title_elem.get_text(strip=True)
                    
                    # Determine impact level
                    impact = 'medium'
                    if element.find(class_=re.compile(r'high|red', re.I)):
                        impact = 'high'
                    elif element.find(class_=re.compile(r'low|green', re.I)):
                        impact = 'low'
                    
                    news_items.append({
                        'title': title,
                        'summary': f"Economic event that may impact gold prices: {title[:100]}",
                        'impact': impact,
                        'source': 'Forex Factory',
                        'timestamp': datetime.now()
                    })
                    
        except Exception as e:
            logger.error(f"Error parsing Forex Factory: {e}")
            
        return news_items

    def _parse_investing_com(self, soup: BeautifulSoup, limit: int) -> List[Dict[str, Any]]:
        """Parse Investing.com gold news"""
        news_items = []
        try:
            # Look for news articles
            articles = soup.find_all(['article', 'div'], class_=re.compile(r'article|news|story'))
            
            for article in articles[:limit]:
                title_elem = article.find(['h1', 'h2', 'h3', 'a'])
                if title_elem:
                    title = title_elem.get_text(strip=True)
                    
                    # Get summary if available
                    summary_elem = article.find(['p', 'div'], class_=re.compile(r'summary|excerpt|description'))
                    summary = summary_elem.get_text(strip=True)[:200] if summary_elem else title[:100]
                    
                    news_items.append({
                        'title': title,
                        'summary': summary,
                        'impact': 'medium',
                        'source': 'Investing.com',
                        'timestamp': datetime.now()
                    })
                    
        except Exception as e:
            logger.error(f"Error parsing Investing.com: {e}")
            
        return news_items

    def _parse_marketwatch(self, soup: BeautifulSoup, limit: int) -> List[Dict[str, Any]]:
        """Parse MarketWatch gold news"""
        news_items = []
        try:
            # Look for news related to gold futures
            news_elements = soup.find_all(['div', 'article'], class_=re.compile(r'news|article|story'))
            
            for element in news_elements[:limit]:
                title_elem = element.find(['h1', 'h2', 'h3', 'a'])
                if title_elem:
                    title = title_elem.get_text(strip=True)
                    
                    # Check if it's gold-related
                    if any(keyword in title.lower() for keyword in ['gold', 'precious', 'metal', 'commodity']):
                        news_items.append({
                            'title': title,
                            'summary': f"MarketWatch gold market update: {title[:100]}",
                            'impact': 'medium',
                            'source': 'MarketWatch',
                            'timestamp': datetime.now()
                        })
                        
        except Exception as e:
            logger.error(f"Error parsing MarketWatch: {e}")
            
        return news_items

    def _parse_fxstreet(self, soup: BeautifulSoup, limit: int) -> List[Dict[str, Any]]:
        """Parse FXStreet gold news"""
        news_items = []
        try:
            # Look for gold analysis and news
            articles = soup.find_all(['article', 'div'], class_=re.compile(r'article|news|analysis'))
            
            for article in articles[:limit]:
                title_elem = article.find(['h1', 'h2', 'h3', 'a'])
                if title_elem:
                    title = title_elem.get_text(strip=True)
                    
                    # Check if it's gold-related
                    if any(keyword in title.lower() for keyword in ['gold', 'xau', 'precious']):
                        news_items.append({
                            'title': title,
                            'summary': f"FXStreet gold analysis: {title[:100]}",
                            'impact': 'medium',
                            'source': 'FXStreet',
                            'timestamp': datetime.now()
                        })
                        
        except Exception as e:
            logger.error(f"Error parsing FXStreet: {e}")
            
        return news_items

    def get_weekly_news_summary(self, limit_per_source: int = 3) -> List[Dict[str, Any]]:
        """Get weekly news summary from all sources"""
        try:
            logger.info("📰 Fetching weekly gold news summary")
            
            all_news = []
            
            # Fetch from each source
            for source_name, url in self.news_sources.items():
                try:
                    news_items = self.fetch_news_from_source(source_name, url, limit_per_source)
                    all_news.extend(news_items)
                    
                    if news_items:
                        logger.info(f"✅ Fetched {len(news_items)} news items from {source_name}")
                    else:
                        logger.warning(f"⚠️ No news items found from {source_name}")
                        
                except Exception as e:
                    logger.error(f"❌ Error fetching from {source_name}: {e}")
                    continue
            
            # Sort by impact and recency
            impact_priority = {'high': 3, 'medium': 2, 'low': 1}
            all_news.sort(key=lambda x: (impact_priority.get(x.get('impact', 'medium'), 2), x.get('timestamp', datetime.min)), reverse=True)
            
            # Return top news items
            top_news = all_news[:10]  # Top 10 news items
            
            logger.info(f"📊 Compiled {len(top_news)} top news items for weekly report")
            return top_news
            
        except Exception as e:
            logger.error(f"❌ Error getting weekly news summary: {e}")
            
            # Return fallback news items if web scraping fails
            return self._get_fallback_news()

    def _get_fallback_news(self) -> List[Dict[str, Any]]:
        """Provide fallback news items when web scraping fails"""
        fallback_news = [
            {
                'title': 'بررسی روند هفتگی قیمت طلا در بازارهای جهانی',
                'summary': 'تحلیل تکنیکال و بنیادی قیمت طلا بر اساس عوامل اقتصادی هفته',
                'impact': 'medium',
                'source': 'تحلیل تکنیکال',
                'timestamp': datetime.now()
            },
            {
                'title': 'نرخ تورم و تأثیر آن بر قیمت فلزات گرانبها',
                'summary': 'بررسی ارتباط نرخ تورم آمریکا با قیمت طلا و نقره',
                'impact': 'high',
                'source': 'تحلیل اقتصادی',
                'timestamp': datetime.now()
            },
            {
                'title': 'تصمیمات بانک مرکزی آمریکا و تأثیر بر بازار طلا',
                'summary': 'بررسی سیاست‌های پولی فدرال رزرو و پیامدهای آن برای طلا',
                'impact': 'high',
                'source': 'تحلیل بانک مرکزی',
                'timestamp': datetime.now()
            },
            {
                'title': 'تنش‌های ژئوپلیتیک و افزایش تقاضا برای پناهگاه امن',
                'summary': 'تأثیر عدم اطمینان سیاسی جهانی بر تقاضای طلا به عنوان پناهگاه امن',
                'impact': 'medium',
                'source': 'تحلیل ژئوپلیتیک',
                'timestamp': datetime.now()
            }
        ]
        
        logger.info("📋 Using fallback news items for weekly report")
        return fallback_news

# Global instance
news_service = NewsService()

def get_weekly_news_summary() -> List[Dict[str, Any]]:
    """Get weekly news summary - global function"""
    return news_service.get_weekly_news_summary()

def get_daily_news() -> List[Dict[str, Any]]:
    """Get daily news - compatibility function"""
    return news_service.get_weekly_news_summary()

def get_news_summary() -> List[Dict[str, Any]]:
    """Get news summary - compatibility function"""
    return news_service.get_weekly_news_summary()[:3]

File: server/services/news-service.ts
import { load } from 'cheerio';

export interface NewsItem {
  id: string;
  title: string;
  description?: string;
  content?: string;
  source: string;
  url?: string;
  publishedAt: Date;
  impact: 'LOW' | 'MEDIUM' | 'HIGH';
  tags: string[];
  relevanceScore?: number;
}

export class NewsService {
  private baseHeaders = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.5',
    'Accept-Encoding': 'gzip, deflate',
    'Connection': 'keep-alive',
    'Upgrade-Insecure-Requests': '1',
  };

  constructor() {
    console.log('📰 NewsService initialized');
  }

  // تایید دریافت فایل CSV دستی از FXStreet
  async confirmManualCSVUpload(fileName: string, recordCount: number): Promise<void> {
    try {
      await storage.createLog({
        level: 'info',
        message: `✅ فایل CSV دستی دریافت شد: ${fileName} با ${recordCount} رکورد`,
        source: 'csv-upload',
        metadata: { fileName, recordCount, uploadTime: new Date().toISOString() }
      });

      console.log(`✅ تایید دریافت فایل CSV: ${fileName} (${recordCount} رویداد)`);
      
      // اطلاع به ادمین
      const telegramModule = await import('../services/telegram');
      await telegramModule.telegramService.sendToAdmin(
        `📊 <b>تایید دریافت فایل CSV</b>\n\n` +
        `📁 نام فایل: ${fileName}\n` +
        `📈 تعداد رکورد: ${recordCount}\n` +
        `⏰ زمان: ${new Date().toLocaleString('fa-IR')}\n\n` +
        `✅ فایل با موفقیت پردازش شد`
      );
    } catch (error) {
      console.error('خطا در تایید فایل CSV:', error);
    }
  }

  async scrapeForexFactory(): Promise<NewsItem[]> {
    try {
      console.log('🔍 Scraping ForexFactory for gold-related news...');
      
      // Add delay to avoid rate limiting
      await new Promise(resolve => setTimeout(resolve, 2000));
      
      const response = await fetch('https://www.forexfactory.com/calendar', {
        headers: this.baseHeaders,
        timeout: 15000
      });

      if (!response.ok) {
        throw new Error(`ForexFactory request failed: ${response.status}`);
      }

      const html = await response.text();
      const $ = load(html);
      const news: NewsItem[] = [];

      // Extract news items from ForexFactory
      $('.calendar__row').each((index, element) => {
        try {
          const titleElement = $(element).find('.calendar__event');
          const impactElement = $(element).find('.calendar__impact');
          const currencyElement = $(element).find('.calendar__currency');
          
          const title = titleElement.text().trim();
          const currency = currencyElement.text().trim();
          const impactSpan = impactElement.find('.calendar__impact-icon').length;
          
          if (title && this.isGoldRelevant(title, currency)) {
            const impact = this.mapImpact(impactSpan);
            
            news.push({
              id: `ff_${Date.now()}_${index}`,
              title: `${currency} ${title}`,
              description: `ForexFactory economic event - Impact: ${impact}`,
              content: title,
              source: 'ForexFactory',
              publishedAt: new Date(),
              impact,
              tags: ['economic', 'forexfactory', 'gold'],
              relevanceScore: this.calculateRelevanceScore(title, currency)
            });
          }
        } catch (itemError) {
          console.error('Error processing ForexFactory item:', itemError);
        }
      });

      console.log(`✅ ForexFactory: Found ${news.length} gold-relevant news items`);
      return news.slice(0, 15); // افزایش محدودیت

    } catch (error) {
      console.error('❌ ForexFactory scraping failed:', error);
      return [];
    }
  }

  async scrapeFXStreet(): Promise<NewsItem[]> {
    try {
      console.log('🔍 Scraping FXStreet for gold analysis...');
      
      const response = await fetch('https://www.fxstreet.com/news', {
        headers: this.baseHeaders,
      });

      if (!response.ok) {
        throw new Error(`FXStreet request failed: ${response.status}`);
      }

      const html = await response.text();
      const $ = load(html);
      const news: NewsItem[] = [];

      // Extract articles from FXStreet
      $('.fxs_article_title_link, .fxs_headline_primary_link').each((index, element) => {
        try {
          const title = $(element).text().trim();
          const url = $(element).attr('href');
          
          if (title && this.isGoldRelevant(title, '')) {
            news.push({
              id: `fxs_${Date.now()}_${index}`,
              title: title,
              description: `FXStreet market analysis`,
              content: title,
              source: 'FXStreet',
              url: url?.startsWith('http') ? url : `https://www.fxstreet.com${url}`,
              publishedAt: new Date(),
              impact: 'MEDIUM',
              tags: ['analysis', 'fxstreet', 'gold'],
              relevanceScore: this.calculateRelevanceScore(title, '')
            });
          }
        } catch (itemError) {
          console.error('Error processing FXStreet item:', itemError);
        }
      });

      console.log(`✅ FXStreet: Found ${news.length} gold-relevant news items`);
      return news.slice(0, 15); // افزایش محدودیت

    } catch (error) {
      console.error('❌ FXStreet scraping failed:', error);
      return [];
    }
  }

  async getAllNews(): Promise<NewsItem[]> {
    try {
      console.log('📰 Fetching news from all sources...');
      
      const [forexFactoryNews, fxStreetNews] = await Promise.all([
        this.scrapeForexFactory(),
        this.scrapeFXStreet()
      ]);

      const allNews = [...forexFactoryNews, ...fxStreetNews];
      
      // Sort by relevance score and date
      allNews.sort((a, b) => {
        if (a.relevanceScore !== b.relevanceScore) {
          return (b.relevanceScore || 0) - (a.relevanceScore || 0);
        }
        return b.publishedAt.getTime() - a.publishedAt.getTime();
      });

      console.log(`📰 Total news collected: ${allNews.length} items`);
      return allNews;

    } catch (error) {
      console.error('❌ Failed to fetch all news:', error);
      return [];
    }
  }

  async getImportantGoldNews(): Promise<NewsItem[]> {
    const allNews = await this.getAllNews();
    
    // فیلتر بهبود یافته برای اخبار مهم
    const importantNews = allNews.filter(news => {
      const score = news.relevanceScore || 0;
      // کاهش threshold برای شناسایی اخبار بیشتر
      return score >= 55 || news.impact === 'HIGH';
    });

    console.log(`📰 Important gold news: ${importantNews.length} items`);
    return importantNews.slice(0, 8); // افزایش تعداد اخبار
  }

  private isGoldRelevant(title: string, currency: string = ''): boolean {
    const text = (title + ' ' + currency).toLowerCase();
    
    // کلمات کلیدی گسترده‌تر برای شناسایی اخبار مرتبط با طلا
    const directGoldKeywords = [
      'gold', 'xau', 'precious metals', 'bullion', 'golden'
    ];
    
    const economicKeywords = [
      'federal reserve', 'fed', 'fomc', 'jerome powell', 'central bank',
      'interest rate', 'rates', 'rate cut', 'rate hike',
      'inflation', 'cpi', 'ppi', 'pce', 'core inflation',
      'employment', 'nonfarm', 'payrolls', 'unemployment', 'jobs',
      'gdp', 'economic growth', 'recession', 'recovery',
      'monetary policy', 'fiscal policy', 'stimulus',
      'dollar', 'usd', 'dxy', 'greenback', 'currency',
      'treasury', 'yield', 'bond', '10-year', 'yields',
      'geopolitics', 'war', 'conflict', 'tensions',
      'china', 'trade war', 'tariffs', 'sanctions',
      'oil', 'crude', 'energy', 'commodities',
      'stock market', 'equity', 'risk-on', 'risk-off',
      'safe haven', 'hedge', 'volatility', 'uncertainty'
    ];
    
    const majorEconomies = [
      'united states', 'usa', 'us ', 'america',
      'european union', 'eu ', 'europe', 'eurozone',
      'china', 'chinese', 'japan', 'japanese',
      'united kingdom', 'uk ', 'britain', 'british'
    ];
    
    // بررسی مستقیم طلا
    if (directGoldKeywords.some(keyword => text.includes(keyword))) {
      return true;
    }
    
    // بررسی شاخص‌های اقتصادی مهم
    const economicMatch = economicKeywords.some(keyword => text.includes(keyword));
    const majorEconomyMatch = majorEconomies.some(keyword => text.includes(keyword));
    
    // بررسی ارزهای مهم
    const importantCurrencies = ['USD', 'EUR', 'GBP', 'JPY', 'CHF', 'CNY'];
    const currencyMatch = importantCurrencies.includes(currency) || 
                         importantCurrencies.some(curr => text.includes(curr.toLowerCase()));
    
    // ترکیب شرایط برای شناسایی بهتر
    return economicMatch || (majorEconomyMatch && currencyMatch);
  }

  private calculateRelevanceScore(title: string, currency: string = ''): number {
    const text = (title + ' ' + currency).toLowerCase();
    let score = 40; // Base score کاهش یافت

    // امتیاز بالا برای کلمات مستقیم طلا
    if (text.includes('gold') || text.includes('xau') || text.includes('bullion')) score += 35;
    if (text.includes('precious metals')) score += 30;
    
    // امتیاز بالا برای Fed و سیاست پولی
    if (text.includes('federal reserve') || text.includes('fed') || text.includes('fomc')) score += 28;
    if (text.includes('jerome powell') || text.includes('powell')) score += 25;
    if (text.includes('interest rate') || text.includes('rates')) score += 25;
    if (text.includes('monetary policy')) score += 22;
    
    // شاخص‌های تورم
    if (text.includes('inflation') || text.includes('cpi') || text.includes('ppi')) score += 20;
    if (text.includes('pce') || text.includes('core inflation')) score += 18;
    
    // داده‌های اشتغال
    if (text.includes('employment') || text.includes('payrolls') || text.includes('nonfarm')) score += 18;
    if (text.includes('unemployment') || text.includes('jobs')) score += 15;
    
    // شاخص‌های اقتصادی
    if (text.includes('gdp') || text.includes('economic growth')) score += 15;
    if (text.includes('recession') || text.includes('recovery')) score += 18;
    
    // دلار و ارز
    if (text.includes('dollar') || text.includes('usd') || text.includes('dxy')) score += 16;
    if (text.includes('currency') || text.includes('greenback')) score += 12;
    
    // اوراق قرضه و yield
    if (text.includes('treasury') || text.includes('yield') || text.includes('bond')) score += 14;
    if (text.includes('10-year') || text.includes('yields')) score += 12;
    
    // ژئوپلیتیک و safe haven
    if (text.includes('geopolitics') || text.includes('war') || text.includes('conflict')) score += 15;
    if (text.includes('safe haven') || text.includes('hedge')) score += 20;
    if (text.includes('uncertainty') || text.includes('volatility')) score += 12;
    
    // کشورهای مهم
    if (text.includes('china') || text.includes('trade war')) score += 12;
    if (text.includes('oil') || text.includes('crude') || text.includes('energy')) score += 10;
    
    // امتیاز ارز
    if (currency === 'USD') score += 12;
    if (['EUR', 'GBP', 'JPY', 'CHF', 'CNY'].includes(currency)) score += 8;
    if (['CAD', 'AUD', 'NZD'].includes(currency)) score += 5;
    
    // بونوس برای ترکیب کلمات
    const combinationBonus = this.getCombinationBonus(text);
    score += combinationBonus;

    return Math.min(score, 100);
  }
  
  private getCombinationBonus(text: string): number {
    let bonus = 0;
    
    // ترکیبات قدرتمند
    if ((text.includes('fed') || text.includes('federal reserve')) && 
        (text.includes('rate') || text.includes('policy'))) bonus += 10;
        
    if (text.includes('inflation') && text.includes('gold')) bonus += 15;
    if (text.includes('dollar') && (text.includes('strength') || text.includes('weak'))) bonus += 8;
    if (text.includes('yield') && text.includes('rise')) bonus += 6;
    if (text.includes('employment') && text.includes('strong')) bonus += 5;
    
    return bonus;
  }

  private mapImpact(impactSpans: number): 'LOW' | 'MEDIUM' | 'HIGH' {
    if (impactSpans >= 3) return 'HIGH';
    if (impactSpans === 2) return 'MEDIUM';
    return 'LOW';
  }

  // Fallback method for when scraping fails
  async getFallbackNews(): Promise<NewsItem[]> {
    console.log('📰 Using fallback news data...');
    
    return [
      {
        id: 'fallback_1',
        title: 'US Federal Reserve Interest Rate Decision Pending',
        description: 'Market awaits Fed decision on interest rates, potential impact on gold prices',
        content: 'The Federal Reserve is expected to announce its interest rate decision, which could significantly impact gold markets.',
        source: 'Market Analysis',
        publishedAt: new Date(),
        impact: 'HIGH',
        tags: ['fed', 'interest-rate', 'gold'],
        relevanceScore: 95
      },
      {
        id: 'fallback_2',
        title: 'US Dollar Strengthens Ahead of Economic Data',
        description: 'Dollar gains could pressure gold prices in short term',
        content: 'The US Dollar has strengthened against major currencies, potentially creating headwinds for gold.',
        source: 'Market Analysis',
        publishedAt: new Date(),
        impact: 'MEDIUM',
        tags: ['usd', 'dollar', 'gold'],
        relevanceScore: 80
      },
      {
        id: 'fallback_3',
        title: 'Inflation Data Shows Mixed Signals',
        description: 'Latest CPI data provides mixed signals for monetary policy',
        content: 'Consumer Price Index data shows mixed signals, creating uncertainty for gold direction.',
        source: 'Market Analysis',
        publishedAt: new Date(),
        impact: 'MEDIUM',
        tags: ['inflation', 'cpi', 'gold'],
        relevanceScore: 75
      }
    ];
  }
}